File: leetcode/20_valid_parentheses.go
func isValid(s string) bool {
	stack := make([]int32, 0, len(s))
	for _, ch := range s {
		switch ch {
		case '(', '[', '{':
			stack = append(stack, ch)
		case ')', ']', '}':
			if len(stack) == 0 {
				return false
			}
			left := stack[len(stack)-1]
			if (left == '(' && ch == ')') || (left == '[' && ch == ']') || (left == '{' && ch == '}') {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		}
	}
	return len(stack) == 0
}

File: leetcode/20_valid_parentheses.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = Vec::<char>::new();
        for c in s.chars() {
            match c {
                '(' | '{' | '[' => {
                    stack.push(c);
                },
                ')' | '}' | ']' => {
                    if stack.is_empty() {
                        return false;
                    }
                    let top = stack[stack.len() - 1];
                    if (c == ')' && top == '(') ||
                        (c == '}' && top == '{') ||
                        (c == ']' && top == '[')
                    {
                        stack.resize(stack.len() - 1, '-');
                    } else { return false; }
                },
                _ => {}
            }
        }
        stack.is_empty()
    }
}